src/Data/Array/NonEmpty.purs

```diff
@@ -361,6 +372,41 @@ fold1 = F.fold1
 intercalate :: forall a. Semigroup a => a -> NonEmptyArray a -> a
 intercalate = F.intercalate
 
+-- | The 'transpose' function transposes the rows and columns of its argument.
+-- | For example,
+-- |
+-- | ```purescript
+-- | transpose 
+-- |   (NonEmptyArray [ NonEmptyArray [1, 2, 3]
+-- |                  , NonEmptyArray [4, 5, 6]
+-- |                  ]) == 
+-- |   (NonEmptyArray [ NonEmptyArray [1, 4]
+-- |                  , NonEmptyArray [2, 5]
+-- |                  , NonEmptyArray [3, 6]
+-- |                  ])
+-- | ```
+-- |
+-- | If some of the rows are shorter than the following rows, their elements are skipped:
+-- |
+-- | ```purescript
+-- | transpose 
+-- |   (NonEmptyArray [ NonEmptyArray [10, 11]
+-- |                  , NonEmptyArray [20]
+-- |                  , NonEmptyArray [30, 31, 32]
+-- |                  ]) == 
+-- |   (NomEmptyArray [ NonEmptyArray [10, 20, 30]
+-- |                  , NonEmptyArray [11, 31]
+-- |                  , NonEmptyArray [32]
+-- |                  ])
+-- | ```
+transpose :: forall a. NonEmptyArray (NonEmptyArray a) -> NonEmptyArray (NonEmptyArray a)
+transpose = unsafeFromArray2D <<< A.transpose <<< toArray2D
```

Comment:
Since `toArray2D` is only used here (and same for `unsafeFromArray2D`), I would define these locally:
```purs
transpose = coerce <<< A.transpose <<< coerce
```

If the above doesn't compile, then I'd write:
```purs
transpose = (coerce :: _ (Array _) -> _ (NonEmptyArray _)) <<< A.transpose <<< (coerce :: _ (NonEmptyArray _) -> _ (Array _))
```
I don't think it's a good idea to expose the two other functions as it might encourage their use. I think such usages should be kept local so we can ensure the variant that the array is non-empty remains true.

Reply:
Another good idea.  The ```coerce``` from ```fromArray2D``` required annotation, the other didn't.